nautobot/extras/jobs.py

```diff
@@ -1205,56 +1148,191 @@ def get_job(class_path, reload=False):
+    Finally, it either returns any data returned from `Job.run()` or re-raises any exception encountered.
     """
-    logger.debug("Running job %s", job_class_path)
 
-    job_class = get_job(job_class_path, reload=True)
-    if job_class is None:
-        raise KeyError(f"Job class not found for class path {job_class_path}")
-    job = job_class()
-    job.request = self.request
-    job_result = JobResult.objects.get(id=self.request.id)
-    payload = {
-        "job_result_id": self.request.id,
-        "job_name": job.name,
-        "user_name": job_result.user.username,
-    }
-    if not job.job_model.has_sensitive_variables:
-        payload["job_kwargs"] = kwargs
+    job, event_payload = _prepare_job(job_class_path, self.request, kwargs)
+
+    result = None
+    status = None
     try:
-        publish_event(topic="nautobot.jobs.job.started", payload=payload)
-        job.before_start(self.request.id, args, kwargs)
-        result = job(*args, **kwargs)
-        job.on_success(result, self.request.id, args, kwargs)
-        job.after_return(JobResultStatusChoices.STATUS_SUCCESS, result, self.request.id, args, kwargs, None)
-        payload["job_output"] = result
-        publish_event(topic="nautobot.jobs.job.completed", payload=payload)
-        return result
+        before_start_result = job.before_start(self.request.id, args, kwargs)
+        if not job._failed:
+            # Call job(), which automatically calls job.run():
+            result = job(*args, **kwargs)
+        else:
+            # don't run the job if before_start() reported a failure, and report the before_start() return value
+            result = before_start_result
+
+        event_payload["job_output"] = result
+        status = JobResultStatusChoices.STATUS_SUCCESS if not job._failed else JobResultStatusChoices.STATUS_FAILURE
+
+        if status == JobResultStatusChoices.STATUS_SUCCESS:
+            job.on_success(result, self.request.id, args, kwargs)
+        else:
+            job.on_failure(result, self.request.id, args, kwargs, None)
+
+        job.after_return(status, result, self.request.id, args, kwargs, None)
+
+        if status == JobResultStatusChoices.STATUS_SUCCESS:
+            return result
+
+        # Report a failure, but with a result rather than an exception and einfo:
+        self.update_state(
+            state=status,
+            meta=result,
+        )
+        # If we return a result, Celery automatically applies STATUS_SUCCESS.
+        # If we raise an exception *other than* `Ignore` or `Reject`, Celery automatically applies STATUS_FAILURE.
+        # We don't want to overwrite the manual state update that we did above, so:
+        raise Ignore()
+
+    except Reject:
+        status = status or JobResultStatusChoices.STATUS_REJECTED
+        raise
+
+    except Ignore:
+        status = status or JobResultStatusChoices.STATUS_IGNORED
+        raise
```

Comment:
In which case would status be `None`? If job.run() raise Ignore or Reject right?

Reply:
Right, if `before_start` or `run` raises either of these exceptions directly (which I understand to be part of the user use case driving this whole body of work) then `status` will not have been set.